### ISSUE TITLE
SGP Extras no longer works on MacOS 12.3 (python removed from OS)

### ISSUE THREAD

**Issue by ZPrimed** [smoothgesturesplus/Issues#81]
Looks like the SGP "Extras" package for MacOS relies on python2, which is no longer packaged with MacOS as of 12.3.

https://www.macrumors.com/2022/01/28/apple-removing-python-2-in-macos-12-3/

Apple suggests bundling python in the app... Personally I have homebrew installed, tried editing the app package to point to the homebrew python installation... but it's python3, not 2.7, so I suspect it's not going to work correctly. (I also had to override the OS because changing main.py in the .app package obviously breaks the signature.)

**Comment by FatherOfAlways** [comment]
Is there a workaround for this? Can you somehow point the app to the correct python directory?
Sorry, not a dev, limited experience with stuff like this.

**Comment by ZPrimed** [comment]
> Is there a workaround for this? Can you somehow point the app to the correct python directory? Sorry, not a dev, limited experience with stuff like this.

Apple does not include *any* version of Python on macOS as of v12.3. They used to include 2.7 in older releases of macOS, but 12.3 has no Python at all. There is a "stub" for `python3` in place, but I believe it only works if you install the Command Line Developer Tools (which are not there by default on macOS, it's an optional add-on). Hence, Apple's suggestion that if your "app" requires a scripting language, it should be included as part of the app bundle.

I actually do have the Developer Tools installed, so I do have `python3` on my Mac. I tried to make the SGP Extras app use python3 instead of python2.  I was able to open up the .app file and see that there are .py files inside. I edited the .py files to point to python3, but things don't seem to be running correctly. I'm not a skilled enough python dev to be able to debug this further, unfortunately (although I suspect it might not be too difficult to fix). The other problem with this work-around is that after opening the .app bundle and editing contents, the "signature" on the app is no longer correct, so macOS will not want to run it unless you right click and choose "Open" (just double-clicking it won't work, because the signature is invalid).

Since this is a paid plug-in, I'm hoping the developer can step in here and get this handled relatively quickly, unless there is some other architectural problem that prevents it from working (I don't know if there have been changes in Chrome that would prevent the "Extras" app from tying in correctly).

**Comment by CharlesMagnuson** [comment]
Just adding a +1 here to note that I too am unable to run SGP due to this issue.

However, I am baffled by the fact that my personal laptop continues to run SGP extras just fine while my work laptop cannot. Personal laptop is running MacOS 12.2.3 while work laptop is running 12.3. They're both running Monterey though, so I don't understand why SGP works on 12.2.3 but not 12.3. I suppose the change that introduced this problem came with v12.3. I sure hope Scott is still working on this project because I am really struggling to do my job without this extension.

**Comment by ZPrimed** [comment]
> I suppose the change that introduced this problem came with v12.3. I sure hope Scott is still working on this project because I am really struggling to do my job without this extension.

Yeah, Apple made this change (that I would consider a relatively *major* change) in a "minor" version update (12.3). Not how I would've done it, were I Apple... save it for v13 or something.

OTOH, they had **ample** warning about it. 

Personally, I moved everything to Vivaldi and am using that, since it has native gesture support baked in...

**Comment by alfredhot** [comment]
> 

OK, after 12.3.1 updating. Even the Extra already installed also cannot run.

**Comment by maxkajiwara** [comment]
I also tried using 2to3 to convert the files, fixed some imports, and changed the path names to python3. At this point I can get it to launch and install/uninstall but it's still not recognized by the extension.

<img width="560" alt="image" src="https://user-images.githubusercontent.com/19397938/184198457-72c20402-a1b1-4856-abda-f800845721e8.png">

**Comment by tofuliang** [comment]
step to get it work under macOS 12.6

1. Download python2 [here](https://www.python.org/ftp/python/2.7.18/python-2.7.18-macosx10.9.pkg) and install.
2. `pip install pyobjc`
3. modify files `PyUserInput/reference_materials/xlib-keysyms-to-python.py`,`sgplus-extras.py`,`main.py` in `/Applications/SmoothGesturesPlusExtras.app/Contents/MacOS`
change first line to `#!/usr/local/bin/python`

**Comment by alfredhot** [comment]
> 
It works!
Thank you ma friend, It's perfect! Good job!

**Comment by alfredhot** [comment]
> step to get it work under macOS 12.6
> 
> 1. Download python2 [here](https://www.python.org/ftp/python/2.7.18/python-2.7.18-macosx10.9.pkg) and install.
> 2. `pip install pyobjc`
> 3. modify files `PyUserInput/reference_materials/xlib-keysyms-to-python.py`,`sgplus-extras.py`,`main.py` in `/Applications/SmoothGesturesPlusExtras.app/Contents/MacOS`
>    change first line to `#!/usr/local/bin/python`

It works!
Thank you ma friend, It's perfect! Good job!

**Comment by Removable** [comment]
> step to get it work under macOS 12.6
> 
> 1. Download python2 [here](https://www.python.org/ftp/python/2.7.18/python-2.7.18-macosx10.9.pkg) and install.
> 2. `pip install pyobjc`
> 3. modify files `PyUserInput/reference_materials/xlib-keysyms-to-python.py`,`sgplus-extras.py`,`main.py` in `/Applications/SmoothGesturesPlusExtras.app/Contents/MacOS`
>    change first line to `#!/usr/local/bin/python`

@tofuliang Hi, I followed your approach, but it doesn't work for me. And the first line of the origin `xlib-keysyms-to-python.py` file is `!#/usr/bin/env python`. Should I change it to what you wrote above or to `!#/usr/local/bin/env python`? Thank anyway!

**Comment by tofuliang** [comment]
> Should I change it to what you wrote above or to !#/usr/local/bin/env python?

@Removable just give it a try and make sure `/usr/local/bin/python` is a symbolic link to python2 like this

```bash
❯ ls -hal /usr/local/bin/python
lrwxr-xr-x 1 root wheel 68  9 21 14:05 /usr/local/bin/python -> ../../../Library/Frameworks/Python.framework/Versions/2.7/bin/python

```

**Comment by shivkaushal** [comment]
> @tofuliang Hi, I followed your approach, but it doesn't work for me. And the first line of the origin `xlib-keysyms-to-python.py` file is `!#/usr/bin/env python`. Should I change it to what you wrote above or to `!#/usr/local/bin/env python`? Thank anyway!

I replaced them with `!#/usr/bin/env python`. 

One thing to look out for is that the instructions is that you might have multiple versions of `pip` installed. It might be worth explicitly running `/usr/bin/local/pip2 install pyobjc` to make sure the right things are getting installed in the right place. This was the thing that got things working for me. 🎉

**Comment by shivkaushal** [comment]
> @tofuliang Hi, I followed your approach, but it doesn't work for me. And the first line of the origin `xlib-keysyms-to-python.py` file is `!#/usr/bin/env python`. Should I change it to what you wrote above or to `!#/usr/local/bin/env python`? Thank anyway!

I also just noticed that your examples were all starting `!#` instead of `#!` - it's important to have this correct 😄

**Comment by klip** [comment]
> step to get it work under macOS 12.6
> 
> 1. Download python2 [here](https://www.python.org/ftp/python/2.7.18/python-2.7.18-macosx10.9.pkg) and install.
> 2. `pip install pyobjc`
> 3. modify files `PyUserInput/reference_materials/xlib-keysyms-to-python.py`,`sgplus-extras.py`,`main.py` in `/Applications/SmoothGesturesPlusExtras.app/Contents/MacOS`
>    change first line to `#!/usr/local/bin/python`

Doesn't work for me.
M1 Pro chip, Mac 12.6

**Comment by Agusum** [comment]
@tofuliang I tried your method but I never get the prompt to enable the permissions for sge (step 5).

I reverted back to the python2 method but now my issue is that I installed the extension in a new laptop and I'm no longer able to verify my license. The button doesn't seem to do anything.

Does anyone know how to fix that?
